subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/SystemDefaultSSLContextFactory.java

```diff
@@ -0,0 +1,169 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resource.transport.http;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import javax.annotation.Nullable;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.security.KeyStore;
+
+/**
+ * This class contains SSLContext initialization similar to what SSLContext.getDefault() does.
+ * Unfortunately, SSLContext.getDefault() can not be used because it is heavily cached by the JVM,
+ * including keyStore and trustStore managers.
+ * There is no way to reset these caches because they are private and are initialized in static sections.
+ */
+@NonNullApi
+public class SystemDefaultSSLContextFactory {
+    private static final String NONE = "NONE";
+    private static final String P11KEYSTORE = "PKCS11";
+
+    private static final Logger LOGGER = Logging.getLogger(SystemDefaultSSLContextFactory.class);
+
+    static SSLContext create() throws Exception {
+        SSLContext context = SSLContext.getInstance("TLS");
+        context.init(getKeyManagers(), getTrustManagers(), null);
+        return context;
+    }
+
+    private static KeyManager[] getKeyManagers() throws Exception {
+        String keyStorePath = System.getProperty("javax.net.ssl.keyStore", "");
+        String keyStoreType = System.getProperty("javax.net.ssl.keyStoreType", KeyStore.getDefaultType());
+        String keyStoreProvider = System.getProperty("javax.net.ssl.keyStoreProvider", "");
+        String keyStorePasswordString = System.getProperty("javax.net.ssl.keyStorePassword", "");
+
+        if (P11KEYSTORE.equals(keyStoreType) && !NONE.equals(keyStorePath)) {
+            throw new IllegalArgumentException("if keyStoreType is " + P11KEYSTORE + ", then keyStore must be " + NONE);
+        }
+
+        char[] keystorePassword = null;
+        if (!keyStorePasswordString.isEmpty()) {
+            keystorePassword = keyStorePasswordString.toCharArray();
+        }
+
+        KeyStore keyStore = loadKeyStore(
+            keyStorePath,
+            keyStoreType,
+            keyStoreProvider,
+            keystorePassword,
+            true
+        );
+
+        KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        if (P11KEYSTORE.equals(keyStoreType)) {
+            keyManagerFactory.init(keyStore, null);
+        } else {
+            keyManagerFactory.init(keyStore, keystorePassword);
+        }
+        return keyManagerFactory.getKeyManagers();
+    }
+
+    private static @Nullable KeyStore loadKeyStore(
```

Comment:
🧐 I've always seen us put the `@Nullable` return type annotation before `private static` access modifiers.

Reply:
My bad, thanks!